multiset.go

```diff
@@ -0,0 +1,97 @@
+package secp256k1
+
+// #include "./depend/secp256k1/include/secp256k1_multiset.h"
+import "C"
+import "github.com/pkg/errors"
+
+// MultiSet is a type used to create an Elliptic Cure Multiset Hash.
+// which is a rolling(homomorphic) hash that you can add and remove elements from
+// and receiving the same resulting hash as-if you never hashed that element.
+// Because of that the order of adding and removing elements doesn't matter.
+// The type should only be created via the supplied methods.
+type MultiSet struct {
+	set C.secp256k1_multiset
+}
+
+// NewMultiset return an empty initialized set.
+// when finalized it should be equal to a finalized set with all elements removed.
+func NewMultiset() MultiSet {
+	multiset := MultiSet{}
+	ret := C.secp256k1_multiset_init(context, &multiset.set)
+	if ret != 1 {
+		panic("failed intializing a multiset. Should never happen")
+	}
+	return multiset
+}
+
+// Reset the multiset. Equivalent to creating a new empty set.
+func (multiset *MultiSet) Reset() {
+	ret := C.secp256k1_multiset_init(context, &multiset.set)
+	if ret != 1 {
+		panic("failed resetting a multiset. Should never happen")
+	}
+}
+
+// Add a data element to the multiset. this will hash the data onto the curve and add it.
+// Supports arbitrary length data(subject to the underlying hash function(SHA256) limits)
+func (multiset *MultiSet) Add(data []byte) {
+	cPtrData := (*C.uchar)(&data[0])
+	CLenData := (C.size_t)(len(data))
+	ret := C.secp256k1_multiset_add(context, &multiset.set, cPtrData, CLenData)
+	if ret != 1 {
+		panic("failed adding to the multiset. Should never happen")
+	}
+}
+
+// Remove a data element from the multiset. this will hash the data onto the curve and remove it.
+// Supports arbitrary length data(subject to the underlying hash function(SHA256) limits)
+func (multiset *MultiSet) Remove(data []byte) {
+	cPtrData := (*C.uchar)(&data[0])
+	CLenData := (C.size_t)(len(data))
+	ret := C.secp256k1_multiset_remove(context, &multiset.set, cPtrData, CLenData)
+	if ret != 1 {
+		panic("failed removing from the multiset. Should never happen")
+	}
+}
+
+// Combine 2 MultiSets together. is equivalent to manually adding all the data elements
+// from one set to the other.
+func (multiset *MultiSet) Combine(input MultiSet) {
+	ret := C.secp256k1_multiset_combine(context, &multiset.set, &input.set)
+	if ret != 1 {
+		panic("failed combining 2 multisets. Should never happen")
+	}
+}
+
+// Finalize the multiset. This will hash the set into a 32 byte hash(SHA256)
+func (multiset *MultiSet) Finalize() [32]byte {
+	hash := [32]byte{}
+	cPtrHash := (*C.uchar)(&hash[0])
+	ret := C.secp256k1_multiset_finalize(context, cPtrHash, &multiset.set)
+	if ret != 1 {
+		panic("failed finalizing the multiset. Should never happen")
+	}
+	return hash
+}
+
+// Serialize the multiset. this is the only right way to serialize a multiset for storage.
```

Comment:
Please add some information regarding the diff between what Finalize returns and Serialize returns - it's not very clear.